src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,6933 @@
+            env(pay(alice, bob, USD(10)),
+                delivermin(Account(carol)["USD"](5)),
+                txflags(tfPartialPayment),
+                ter(temBAD_AMOUNT));
+            env(pay(alice, bob, USD(10)),
+                delivermin(USD(15)),
+                txflags(tfPartialPayment),
+                ter(temBAD_AMOUNT));
+            env(pay(gw, carol, USD(50)));
+            AMM ammCarol(env, carol, XRP(10), USD(15));
+            env(pay(alice, bob, USD(10)),
+                paths(XRP),
+                delivermin(USD(7)),
+                txflags(tfPartialPayment),
+                sendmax(XRP(5)),
+                ter(tecPATH_PARTIAL));
+            env.require(balance(alice, XRP(9999.99999)));
+            env.require(balance(bob, XRP(10000)));
+        }
+
+        {
+            Env env(*this, features);
+            fund(env, gw, {alice, bob}, XRP(10000));
+            env.trust(USD(1100), alice, bob);
+            env(pay(gw, bob, USD(1100)));
+            AMM ammBob(env, bob, XRP(1000), USD(1100));
+            env(pay(alice, alice, USD(10000)),
+                paths(XRP),
+                delivermin(USD(100)),
+                txflags(tfPartialPayment),
+                sendmax(XRP(100)));
+            env.require(balance(alice, USD(100)));
+        }
+
+        {
+            Env env(*this, features);
+            fund(env, gw, {alice, bob, carol}, XRP(10000));
+            env.trust(USD(1200), bob, carol);
+            env(pay(gw, bob, USD(1200)));
+            AMM ammBob(env, bob, XRP(5500), USD(1200));
+            env(pay(alice, carol, USD(10000)),
+                paths(XRP),
+                delivermin(USD(200)),
+                txflags(tfPartialPayment),
+                sendmax(XRP(1000)),
+                ter(tecPATH_PARTIAL));
+            env(pay(alice, carol, USD(10000)),
+                paths(XRP),
+                delivermin(USD(200)),
+                txflags(tfPartialPayment),
+                sendmax(XRP(1100)));
+            BEAST_EXPECT(
+                ammBob.expectBalances(XRP(6600), USD(1000), ammBob.tokens()));
+            env.require(balance(carol, USD(200)));
+        }
+
+        {
+            Env env(*this, features);
+            fund(env, gw, {alice, bob, carol, dan}, XRP(10000));
+            env.trust(USD(1100), bob, carol, dan);
+            env(pay(gw, bob, USD(100)));
+            env(pay(gw, dan, USD(1100)));
+            env(offer(bob, XRP(100), USD(100)));
+            env(offer(bob, XRP(1000), USD(100)));
+            AMM ammDan(env, dan, XRP(1000), USD(1100));
+            env(pay(alice, carol, USD(10000)),
+                paths(XRP),
+                delivermin(USD(200)),
+                txflags(tfPartialPayment),
+                sendmax(XRP(200)));
+            env.require(balance(bob, USD(0)));
+            env.require(balance(carol, USD(200)));
+            BEAST_EXPECT(
+                ammDan.expectBalances(XRP(1100), USD(1000), ammDan.tokens()));
```

Comment:
It may not be worth it this point, but I have noticed so many instances of this pattern where the `lpt` parameter to `expectBalances` is just the same AMM's `tokens()`, but maybe change the function so that `lpt` is a `std::optional<IOUAmount>`, then check `lptAMMBalance == STAmount(lpt? *lpt : tokens(), lptIssue_)`, then remove the unnecessary parameter for all these calls....